GPT2 with one step beam search

xi-liu-ds opened the pull request:
**Description**:  In this PR we add support for one-step beam search in ONNX graph.

**Motivation and Context**
- Language models (e.g. GPT) usually need beam search to decode output token sequences. This PR will fulfill this purpose by adding GPT2LMHeadModel_BeamSearchStep class and one-step beam search included directly into ONNX compute graph.

**How to run**
In bash/CMD/PowerShell, run:

```
python onnxruntime/python/tools/transformers/convert_to_onnx.py \
--model_name_or_path=[path/to/model/folder] --model_class=GPT2LMHeadModel_BeamSearchStep --output=[path/to/output/onnx_file_name] -o --precision=int8
```

and optionally with `--input_test_file=[path/to/test/file]`

tianleiwu commented on onnxruntime/python/tools/transformers/gpt2_beamsearch_helper.py:
```diff
@@ -0,0 +1,822 @@
+        return w
+
+class GPT2LMHeadModel_BeamSearchStep(GPT2LMHeadModel):
+    """Here we wrap a class for Onnx model conversion for GPT2LMHeadModel with past state."""
```
It is better to indicate that it includes one step beam search.

xi-liu-ds replied:
Thanks for the suggestion. Added in the update.

tianleiwu commented on onnxruntime/python/tools/transformers/gpt2_beamsearch_tester.py:
```diff
@@ -0,0 +1,435 @@
+        self.last_state = (
+            torch.from_numpy(output[0])
+            if isinstance(output[0], numpy.ndarray)
+            else output[0].clone().detach().cpu()
```
Here the output is copied to CPU. Did you run test in GPU?

xi-liu-ds replied:
Our one step beam search scenario is on CPU only. We didn't test it on GPU specifically.

I saw it passed GPU CI pipeline check. I can do a test on GPU at our side.

xi-liu-ds replied:
I have pushed a few changes and made test. Now one step beam search works on GPU with fp16 model

tianleiwu commented on onnxruntime/python/tools/transformers/gpt2_beamsearch_helper.py:
```diff
@@ -0,0 +1,822 @@
+    ):
+        super().__init__(input_ids, position_ids, attention_mask, past)
+        self.prev_step_results: torch.LongTensor = prev_step_results
+        self.prev_step_scores: torch.FloatTensor = prev_step_scores
```
Union[torch.FloatTensor, torch.HalfTensor]? Is the input always in CPU?

When model is fp16, the data type could be torch.HalfTensor. 
If the input is in GPU, it could also be torch.cuda.FloatTensor.

xi-liu-ds replied:
Updated. 

The input will always be in CPU.

xi-liu-ds replied:
I have pushed a few changes. Now the input works on GPU as well.

tianleiwu commented on onnxruntime/python/tools/transformers/gpt2_beamsearch_helper.py:
```diff
@@ -0,0 +1,822 @@
+                self.input_ids,
+                self.position_ids,
+                self.attention_mask,
+                self.past,
```
past could be the last element so that it has same order as to_list().

xi-liu-ds replied:
Good catch! This function is removed in the update due to it is never used

tianleiwu commented on onnxruntime/python/tools/transformers/gpt2_beamsearch_helper.py:
```diff
@@ -0,0 +1,822 @@
+            gpt2_inputs.attention_mask,
+            gpt2_inputs.past,
+            self.beam_select_idx,
+            self.input_log_probs,
```
Shall input_log_probs etc be converted to fp32 (if it could be fp16)?

xi-liu-ds replied:
Actually input_log_probs etc can't be fp16

xi-liu-ds replied:
> Shall input_log_probs etc be converted to fp32 (if it could be fp16)?

Sorry I misunderstood. Good catch! I have made changes.

hariharans29 commented:
Hi @xi-liu-ds - Thanks for the PR and contribution. Do you have some time to address the comments in the PR ?

hariharans29 commented:
/azp run Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, centos7_cpu, centos7_cpu (linux_centos_ci Debug), centos7_cpu (linux_centos_ci Release), orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-amd-gpu-ci-pipeline, Linux Nuphar CI Pipeline, orttraining-distributed

hariharans29 commented:
/azp run Linux CPU CI Pipeline, Linux CPU Minimal Build E2E CI Pipeline, Linux CPU x64 NoContribops CI Pipeline, Linux GPU TensorRT CI Pipeline, Linux OpenVINO CI Pipeline, MacOS CI Pipeline, MacOS NoContribops CI Pipeline, Windows CPU CI Pipeline, orttraining-ortmodule, orttraining-ortmodule-distributed

hariharans29 commented:
/azp run Linux GPU CI Pipeline

xi-liu-ds commented:
> Hi @xi-liu-ds - Thanks for the PR and contribution. Do you have some time to address the comments in the PR ?

Hi @hariharans29 sorry I was busy with another task. I am working on addressing the comments right now. 

BTW, I am wondering can I kick off all pipelines myself or I have to go through you guys?

Thanks!

hariharans29 commented:
> > Hi @xi-liu-ds - Thanks for the PR and contribution. Do you have some time to address the comments in the PR ?
> 
> Hi @hariharans29 sorry I was busy with another task. I am working on addressing the comments right now.
> 
> BTW, I am wondering can I kick off all pipelines myself or I have to go through you guys?
> 
> Thanks!

Thank you for addressing the comments. We will have to kick off pipelines for you. Let me know when you are done with addressing the comments and I ll kick them off. Thanks.

xi-liu-ds commented:
> > > Hi @xi-liu-ds - Thanks for the PR and contribution. Do you have some time to address the comments in the PR ?
> > 
> > 
> > Hi @hariharans29 sorry I was busy with another task. I am working on addressing the comments right now.
> > BTW, I am wondering can I kick off all pipelines myself or I have to go through you guys?
> > Thanks!
> 
> Thank you for addressing the comments. We will have to kick off pipelines for you. Let me know when you are done with addressing the comments and I ll kick them off. Thanks.

Hi @tianleiwu @hariharans29  I have addressed all the comments. Now it is ready for another round of review. @hariharans29 Could you help kick off pipelines? Thank you all!

hariharans29 commented:
/azp run Windows GPU CI Pipeline, Windows GPU TensorRT CI Pipeline, centos7_cpu, centos7_cpu (linux_centos_ci Debug), centos7_cpu (linux_centos_ci Release), orttraining-linux-ci-pipeline, orttraining-linux-gpu-ci-pipeline, orttraining-amd-gpu-ci-pipeline, Linux Nuphar CI Pipeline, orttraining-distributed
